prl-tools: 12.2.1-41615 -> 17.1.4-51567

catap opened the pull request:
###### Description of changes

Update prl-tools to the last version, tested that it works fine under Parallels 17.

It also:
 - fixes https://github.com/NixOS/nixpkgs/issues/179117,
 - fixes https://github.com/NixOS/nixpkgs/issues/31882,
 - and fixes https://github.com/NixOS/nixpkgs/issues/79280

This PR + https://github.com/NixOS/nixpkgs/pull/179487 and simple overlays:
```nix
    self: super: {
      systemd = super.systemd.overrideAttrs (old: {
        patches = old.patches ++ [
          (super.fetchpatch {
            url = "https://gist.githubusercontent.com/catap/be19980b57e01015ada609e448ca41b8/raw/ba7e61a07bddd59336a0519dfabed2551b5a0cbc/virt-fix-detection-of-Parallels-virtualization.patch";
            sha256 = "sha256-PyCUgioN3P/C+SzwYXLaIdsWuWyT1jJvYsTPFypASyw=";
          })
        ];
      });
    }
```

allows to install parallels tools on VM fully automatically :)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

SuperSandro2000 commented on pkgs/os-specific/linux/prl-tools/default.nix:
```diff
@@ -57,112 +63,64 @@ stdenv.mkDerivation rec {
     description = "Parallels Tools for Linux guests";
     homepage = "https://parallels.com";
-    platforms = [ "i686-linux" "x86_64-linux" ];
+    platforms = lib.platforms.linux;
```
```suggestion
    platforms = platforms.linux;
```

catap replied:
it won't work without `with lib` I guess

catap replied:
=> removed all `lib.`

SuperSandro2000 replied:
we have with lib so this works

SuperSandro2000 commented on pkgs/os-specific/linux/prl-tools/default.nix:
```diff
@@ -1,49 +1,55 @@
+  aarch64 = stdenv.hostPlatform.system == "aarch64-linux";
+  x86_64 = stdenv.hostPlatform.system == "x86_64-linux";
+  i686 = stdenv.hostPlatform.system == "i686-linux";
+  _ = if (aarch64 || x86_64 || i686) then true
+      else throw "Parallels Tools for Linux only supports {aarch64,x86_64,i686}-linux targets";
```
Please use assert instead.

SuperSandro2000 reviewed (commented):
Please squash the commits together for the package.

sandangel reviewed (approved):
thanks a lot for working on this

flokli commented on nixos/modules/virtualisation/parallels-guest.nix:
```diff
@@ -4,6 +4,7 @@ with lib;
 
 let
   prl-tools = config.hardware.parallels.package;
+  aarch64 = pkgs.stdenv.hostPlatform.system == "aarch64-linux";
```
This is only used in one place. Please inline there.

flokli commented on nixos/modules/virtualisation/parallels-guest.nix:
```diff
@@ -44,50 +46,31 @@ in
     boot.extraModulePackages = [ prl-tools ];
 
-    boot.kernelModules = [ "prl_tg" "prl_eth" "prl_fs" "prl_fs_freeze" ];
+    boot.kernelModules = [ "prl_fs" "prl_fs_freeze" "prl_tg" ] ++ optional aarch64 "prl_notifier";
```
please inline the aarch64 check here (and break this up in two lines)

SuperSandro2000 commented on nixos/modules/virtualisation/parallels-guest.nix:
```diff
@@ -44,50 +45,32 @@ in
 
-    boot.kernelModules = [ "prl_tg" "prl_eth" "prl_fs" "prl_fs_freeze" ];
+    boot.kernelModules = [ "prl_fs" "prl_fs_freeze" "prl_tg" ]
+                         ++ optional (pkgs.stdenv.hostPlatform.system == "aarch64-linux") "prl_notifier";
```
```suggestion
      ++ optional (pkgs.stdenv.hostPlatform.system == "aarch64-linux") "prl_notifier";
```

catap commented:
@SuperSandro2000 may I ask you re-review one more time?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/547

SuperSandro2000 commented on pkgs/os-specific/linux/prl-tools/default.nix:
```diff
@@ -1,49 +1,49 @@
 stdenv.mkDerivation rec {
-  version = "${prl_major}.2.1-41615";
-  prl_major = "12";
+  version = "${prl_major}.1.4-51567";
+  prl_major = "17";
```
```suggestion
  version = "17.1.4-51567";
```

We can get prl_major like this:
```
nix-repl> lib.versions.major "17.1.4-51567"
"17"
```

catap replied:
@SuperSandro2000 thanks, fixed.

SuperSandro2000 commented:
> @SuperSandro2000 may I ask you re-review one more time?

I don't really know anything about prl-tools so I am a bit hesitant with merging.

catap commented:
@SuperSandro2000 it is working. I'm using it for a few weeks for now.

I assume that @wegank and @sandangel also had tested this code and enjoy it.

How can I help you to decide to merge it?

negz commented:
FWIW I've been running this for a week per https://github.com/negz/nix/blob/be24cbb/hosts/mael/hardware-configuration.nix#L7.

The only issue I've noticed so far is that `prltimesync` isn't running - I notice if I start a Fedora VM that it's a child of `prltoolsd` along with a few other things. On my system `prltoolsd` has no children:

```
$ pstree 1460373
--= 1460373 root /nix/store/zbw072dym5mlnjg26zz1xk4cxsm6dr7g-prl-tools-17.1.4-51567/bin/prltoolsd -f

$ systemctl status prltoolsd.service
● prltoolsd.service - Parallels Tools Service
     Loaded: loaded (/etc/systemd/system/prltoolsd.service; enabled; vendor preset: enabled)
     Active: active (running) since Fri 2022-07-08 21:20:25 PDT; 1min 19s ago
   Main PID: 1460373 (prltoolsd)
         IP: 0B in, 0B out
         IO: 0B read, 4.0K written
      Tasks: 4 (limit: 19088)
     Memory: 392.0K
        CPU: 205ms
     CGroup: /system.slice/prltoolsd.service
             └─1460373 /nix/store/zbw072dym5mlnjg26zz1xk4cxsm6dr7g-prl-tools-17.1.4-51567/bin/prltoolsd -f

Jul 08 21:20:25 mael systemd[1]: Started Parallels Tools Service.
```

Also worth noting that I'm running without a graphical session so I'm not using/testing any wanted by `wantedBy = [ "graphical-session.target" ];`.

catap commented:
@negz interesting, digging why.

catap commented:
It has much more hardcoded paths. Will try to patch everything in a few days.

catap commented:
@negz, @wegank can I ask you to check one more time?

Right now it keeps `prltimesync` running and also printers should work.

The only error / missed part that I found is: ` Error: prlusmd account manager error when find sessions type: can't open dir` but I have no idea how `prlusmd` should be used in terms of nix, do you?

catap commented:
Just to confirm that copy-and-paste and changes screen size is working. Or at least at xorg + pantheon.

catap committed:
prl-tools: 12.2.1-41615 -> 17.1.4-51567

Install Parallel Tools updated for version 17 of Parallels for macOS. This
fixes clipboard sharing, so that copy and paste works between the host
macOS and the guest NixOS VM. Support for guests on M1 Apple Silicon-based
Macs (aarch64-linux) is also added.

Co-authored-by: Paul Smith <paulsmith@gmail.com>
Co-authored-by: Weijia Wang <9713184+wegank@users.noreply.github.com>

negz commented:
@catap Thank you! I copied your changes over per https://github.com/negz/nix/commit/a0d9a599 and now I see `prltimesync` is a child of `prltoolsd`. FWIW I see `prlusmd` running without any errors, but I have no idea what it does.

```console
$ systemctl status prltoolsd.service
● prltoolsd.service - Parallels Tools Service
     Loaded: loaded (/etc/systemd/system/prltoolsd.service; enabled; vendor preset: enabled)
     Active: active (running) since Sat 2022-07-09 14:55:03 PDT; 1min 54s ago
   Main PID: 1725899 (.prltoolsd-wrap)
         IP: 700B in, 980B out
         IO: 2.4M read, 4.0K written
      Tasks: 9 (limit: 19088)
     Memory: 3.9M
        CPU: 420ms
     CGroup: /system.slice/prltoolsd.service
             ├─1725899 /nix/store/lv1qnh1fwd7py5x7qx2nhyawnadphfin-prl-tools-17.1.4-51567/bin/prltoolsd -f
             ├─1725915 ./prltimesync
             └─1725916 ./prlusmd
```

Also FYI I'm running this on an M1 Mac so I can confirm it's all working (at least the non-graphical stuff) on Linux aarch64:

```console
$ uname -a
Linux mael 5.18.7 #1-NixOS SMP Sat Jun 25 13:29:48 UTC 2022 aarch64 GNU/Linux
```

sandangel commented:
> I don't really know anything about prl-tools so I am a bit hesitant with merging.

hi @SuperSandro2000 , I think it's impossible to know everything when maintaining this huge repository. What can we do to make you feel more confident before merging? What are the risks that you can think of if we merge this PR? Should we add some kind of testing?

catap commented:
@sandangel to be honest I have no idea how to test this software outside of parallels environment :)

SuperSandro2000 commented:
> What are the risks that you can think of if we merge this PR?

The worst that can happen is that people write angry comments.

Testing for unfree packages is a bit tricky. Maybe we just merge this and go from there if something breaks.

ylluminate commented:
Has anyone done any work with this on Parallels on Apple Silicon?

sandangel commented:
I'm using this on M1 mac, work perfectly fine.

wegank commented on nixos/modules/virtualisation/parallels-guest.nix:
```diff
@@ -44,109 +45,99 @@ in
       bindsTo = [ "cups.service" ];
+      path = [ prl-tools ];
       serviceConfig = {
         Type = "forking";
```
TODO: recheck

wegank replied:
TODO: recheck

sandangel commented:
btw, is there a way to use the module and package from unstable, but still everything else under `release-22.05` ?

catap commented:
> btw, is there a way to use the module and package from unstable, but still everything else under `release-22.05` ?

Nope, you can’t overlay a module.

sandangel commented:
@SuperSandro2000 , hi, could you share the code? currently I already copied the module code into local folder. but I would like to remove that and somehow pull it from nix unstable branch.

SuperSandro2000 commented:
I think that should help you https://discourse.nixos.org/t/overriding-a-module/6464/3?u=sandro

hgl commented:
@catap do you use Wayland? I tried this out for a Wayland setup using Parallels 17.1.4-51567 on an Intel Mac, encountered some issues. 

I tried two configurations:

### 1. Wayland with gdm as the display manager:

```nix
{ config, pkgs, ... }:

{
  imports = [ ./hardware-configuration.nix ];
  boot.loader.systemd-boot.enable = true;
  boot.loader.efi.canTouchEfiVariables = true;
  hardware.parallels.enable = true;
  networking.useDHCP = false;
  networking.interfaces.enp0s5.useDHCP = true;

  services.xserver = {
    enable = true;
    displayManager.gdm.enable = true;
    displayManager.autoLogin.enable = true;
    displayManager.autoLogin.user = "hgl";
  };
  programs.sway = {
    enable = true;
    extraSessionCommands = ''
      export WLR_NO_HARDWARE_CURSORS=1
    '';
  };

  users.mutableUsers = false;
  users.users.hgl = {
    isNormalUser = true;
    extraGroups = [ "wheel" ];
    hashedPassword = "<password>";
  };

  system.stateVersion = "22.11";
}
```
With this config, the mouse is very sluggish, trailing behind my real movement.

### 2. Wayland with greetd as the display manager:

```nix
{ config, pkgs, ... }:

{
  imports = [ ./hardware-configuration.nix ];
  boot.loader.systemd-boot.enable = true;
  boot.loader.efi.canTouchEfiVariables = true;
  hardware.parallels.enable = true;
  networking.useDHCP = false;
  networking.interfaces.enp0s5.useDHCP = true;

  services.greetd = {
    enable = true;
    settings = rec {
      initial_session = {
        command = "${pkgs.sway}/bin/sway";
        user = "hgl";
      };
      default_session = initial_session;
    };
  };
  programs.sway = {
    enable = true;
    extraSessionCommands = ''
      export WLR_NO_HARDWARE_CURSORS=1
    '';
  };

  users.mutableUsers = false;
  users.users.hgl = {
    isNormalUser = true;
    extraGroups = [ "wheel" ];
    hashedPassword = "<password>";
  };

  system.stateVersion = "22.11";
}
```
With this config, whenever my mouse moves into the VM window, the mouse disappears.

And in both cases, the clipboard is not shared between the host and guest.

Maybe I missed setting something?

catap commented:
@hgl nope, I haven't use / try it :(

hgl commented:
@sandangel Thanks. Switching to gnome did make all the issues go away.
